File: frontend/src/types/analysis.ts
export type MarketDirection = "bullish" | "bearish" | "neutral";
export type AnalysisSentimentLabel = "positive" | "negative" | "neutral";

export interface MarketPulse {
  value: number;
  direction: MarketDirection;
  sample_size: number;
}

export interface DirectionDistributionItem {
  direction: MarketDirection;
  count: number;
  ratio: number;
}

export interface StockRankingItem {
  stock_id: number;
  ticker: string | null;
  signal_count: number;
  avg_impact_score: number;
  max_abs_impact_score: number;
  dominant_direction: MarketDirection;
}

export interface LatestEventItem {
  article_id: number;
  article_title: string;
  published_at: string;
  source_name: string | null;
  stock_id: number;
  ticker: string | null;
  direction: MarketDirection;
  impact_score: number;
  confidence: number;
}

export interface MarketAnalysisData {
  generated_at: string;
  market_pulse: MarketPulse;
  direction_distribution: DirectionDistributionItem[];
  stock_rankings: StockRankingItem[];
  latest_events: LatestEventItem[];
}

export type MarketAnalysisQueryParams = Record<
  "window_hours" | "article_limit" | "top_stocks" | "latest_events",
  number
>;

export interface OverviewStatsCards {
  total_articles: number;
  analyzed_sentiment_articles: number;
  bullish_signals: number;
  bearish_signals: number;
}

export interface SentimentDistributionItem {
  label: AnalysisSentimentLabel;
  count: number;
  ratio: number;
}

export interface TopImpactSignalItem {
  article_id: number;
  article_title: string;
  published_at: string;
  stock_id: number;
  ticker: string | null;
  direction: MarketDirection;
  impact_score: number;
  confidence: number;
}

export interface AnalysisOverviewData {
  generated_at: string;
  stats_cards: OverviewStatsCards;
  sentiment_distribution: SentimentDistributionItem[];
  top_impact_signals: TopImpactSignalItem[];
}

export type AnalysisOverviewQueryParams = Record<"window_hours" | "article_limit" | "top_signals", number>;

File: backend/app/schemas/analysis.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.common import ApiResponse


class SentimentAggregate(BaseModel):
    label: Literal["positive", "negative", "neutral"]
    score: float


class ImpactItem(BaseModel):
    stock_id: int
    direction: Literal["bullish", "bearish", "neutral"] | str
    score: float
    confidence: float

    model_config = ConfigDict(from_attributes=True)


class MarketPulse(BaseModel):
    value: float
    direction: Literal["bullish", "bearish", "neutral"]
    sample_size: int


class DirectionShare(BaseModel):
    direction: Literal["bullish", "bearish", "neutral"]
    count: int
    ratio: float


class StockRankingItem(BaseModel):
    stock_id: int
    ticker: str | None = None
    signal_count: int
    avg_impact_score: float
    max_abs_impact_score: float
    dominant_direction: Literal["bullish", "bearish", "neutral"]


class LatestEventItem(BaseModel):
    article_id: int
    article_title: str
    published_at: datetime
    source_name: str | None = None
    stock_id: int
    ticker: str | None = None
    direction: Literal["bullish", "bearish", "neutral"] | str
    impact_score: float
    confidence: float


class MarketAnalysisData(BaseModel):
    generated_at: datetime
    market_pulse: MarketPulse
    direction_distribution: list[DirectionShare]
    stock_rankings: list[StockRankingItem]
    latest_events: list[LatestEventItem]


class MarketAnalysisResponse(ApiResponse[MarketAnalysisData]):
    pass


class OverviewStatsCards(BaseModel):
    total_articles: int
    analyzed_sentiment_articles: int
    bullish_signals: int
    bearish_signals: int


class SentimentDistributionItem(BaseModel):
    label: Literal["positive", "negative", "neutral"]
    count: int
    ratio: float


class TopImpactSignalItem(BaseModel):
    article_id: int
    article_title: str
    published_at: datetime
    stock_id: int
    ticker: str | None = None
    direction: Literal["bullish", "bearish", "neutral"]
    impact_score: float
    confidence: float


class AnalysisOverviewData(BaseModel):
    generated_at: datetime
    stats_cards: OverviewStatsCards
    sentiment_distribution: list[SentimentDistributionItem]
    top_impact_signals: list[TopImpactSignalItem]


class AnalysisOverviewResponse(ApiResponse[AnalysisOverviewData]):
    pass
